src/token_retriever/l1_retriever.rs

```diff
@@ -0,0 +1,142 @@
+use std::fmt;
+
+use http::{header::CONTENT_TYPE, StatusCode, Uri};
+use serde_json::json;
+
+use crate::{
+    authenticator::TokenRetrievalResponse,
+    http_client::{HttpClient, HttpClientError},
+    system_identity::input_data::auth_method::ClientSecret,
+    token::Token,
+    TokenRetriever, TokenRetrieverError,
+};
+
+/// HTTP-based token retriever for L1 authentication method (client ID + client secret)
+pub struct L1TokenRetriever<C: HttpClient> {
+    client_id: String,
+    client_secret: ClientSecret,
+    http_client: C,
+    token_retrieval_uri: Uri,
+}
+
+impl<C: HttpClient> fmt::Debug for L1TokenRetriever<C> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("L1TokenRetriever")
+            .field("client_id", &self.client_id)
+            .field("client_secret", &"<hidden>") // Do not print the secret
+            .field("token_retrieval_uri", &self.token_retrieval_uri)
+            .field("http_client", &"impl HttpClient") // HttpClient does not implement Debug
+            .finish()
+    }
+}
+
+impl<C: HttpClient> L1TokenRetriever<C> {
+    pub(super) fn new(
+        client_id: String,
+        client_secret: ClientSecret,
+        http_client: C,
+        token_retrieval_uri: Uri,
+    ) -> Self {
+        Self {
+            client_id,
+            client_secret,
+            http_client,
+            token_retrieval_uri,
+        }
+    }
+
+    fn build_request(
+        client_id: &str,
+        client_secret: &ClientSecret,
+        uri: &Uri,
+    ) -> Result<http::Request<Vec<u8>>, TokenRetrieverError> {
+        let json_body_string = json!({
+            "client_id": client_id,
+            "client_secret": client_secret,
+            "grant_type": "client_credentials",
+        });
+        let json_body = serde_json::to_vec(&json_body_string).map_err(|e| {
+            TokenRetrieverError::TokenRetrieverError(format!("Failed to encode JSON: {e}"))
+        })?;
+
+        http::Request::builder()
+            .uri(uri)
+            .method("POST")
+            .header(CONTENT_TYPE, "application/json")
+            .body(json_body)
+            .map_err(|e| {
+                TokenRetrieverError::TokenRetrieverError(format!("Failed to build request: {e}"))
+            })
+    }
+}
+
+impl<C: HttpClient> TokenRetriever for L1TokenRetriever<C> {
+    fn retrieve(&self) -> Result<Token, TokenRetrieverError> {
```

Comment:
isn't this token retreiver duplicating the Authenticator code? meaning that is also forming a SI service request based on a SI type and obtaining a Token based on that